torch/csrc/jit/type.h

```diff
@@ -183,22 +248,22 @@ struct TORCH_API TensorType : public Type {
   }
   static TypePtr fromNumberType(TypePtr typ);
 
+  static CompleteTensorTypePtr sliceSubtypes(const CompleteTensorTypePtr& type) {
```

Comment:
What does this do? When would I use it?

Reply:
I think I'll slightly change this part before we merge it. There's a subtlety in when you do `type_ptr->cast<TensorType>()`, because if it's a `CompleteTensorType`, it will give you a `TensorTypePtr`. Now, if you later do `node->output()->setType(type->expect<TensorType>())`, then your code probably assumes that you don't care about the sizes, but in reality you never sliced the derived object to the `TensorType` superclass, and so `node->output()` will also inherit its shapes and strides, which might not be valid.

I discovered this because it caused a bug in shape prop. I had a `chunk` with an input with `CompleteTensorType`, it didn't have a case in the complete-shape case, but it did fit into the simplified function, where it simply did
```cpp
std::vector<TensorTypePtr> tensor_types;
...
for (Value * output : node->outputs()) {
  output->setType(tensor_types.at(0));
}
```
thus inheriting pre-chunk sizes in the output, which is obviously incorrect.

Reply:
Regarding the change I want to make: I want to `->cast<TensorType>()` to always slice the object now, so we can prevent those kinds of errors in the future.

Reply:
Yeah, this is why my default assumption is that you shouldn't use subclasses in your design, unless you are ABSOLUTELY CERTAIN that you need the subtyping. In this case, just having CompleteTensorType hold a TensorType inside would have also worked.